File: app/src/main/java/net/emilla/math/BinaryOperator.java
package net.emilla.math;

import androidx.annotation.Nullable;

import net.emilla.math.CalcToken.InfixToken;

enum BinaryOperator implements InfixToken {
    PLUS(1, false) {
        @Override
        public double apply(double a, double b) {
            return a + b;
        }
    },
    MINUS(1, false) {
        @Override
        public double apply(double a, double b) {
            return a - b;
        }
    },
    TIMES(2, false) {
        @Override
        public double apply(double a, double b) {
            return a * b;
        }
    },
    DIV(2, false) {
        @Override
        public double apply(double a, double b) {
            return a / b;
        }
    },
    POW(3, true) {
        @Override
        public double apply(double a, double b) {
            return Math.pow(a, b);
        }
    };

    public abstract double apply(double a, double b);

    @Nullable
    public static final BinaryOperator LPAREN = null;

    public final int precedence;
    public final boolean rightAssociative;

    BinaryOperator(int precedence, boolean rightAssociative) {
        this.precedence = precedence;
        this.rightAssociative = rightAssociative;
    }

    public static BinaryOperator of(char token) {
        return switch (token) {
            case '+' -> PLUS;
            case '-' -> MINUS;
            case '*' -> TIMES;
            case '/' -> DIV;
            case '^' -> POW;
            default -> throw new IllegalArgumentException("Invalid binary operator");
        };
    }

}

File: app/src/main/java/net/emilla/math/BinaryOperator.kt
package net.emilla.math;

/*internal*/ enum BinaryOperator implements InfixToken {
    ADD(1, false) {
        @Override
        double apply(double a, double b) {
            return a + b;
        }
    },
    SUBTRACT(1, false) {
        @Override
        double apply(double a, double b) {
            return a - b;
        }
    },
    TIMES(2, false) {
        @Override
        double apply(double a, double b) {
            return a * b;
        }
    },
    DIV(2, false) {
        @Override
        double apply(double a, double b) {
            return a / b;
        }
    },
    POW(3, true) {
        @Override
        double apply(double a, double b) {
            return Math.pow(a, b);
        }
    };

    static BinaryOperator of(char token) {
        // todo: nat-language words like "add", "to the power of", ..
        return switch (token) {
            case '+' -> ADD;
            case '-' -> SUBTRACT;
            case '*' -> TIMES;
            case '/' -> DIV;
            case '^' -> POW;
            default -> throw new IllegalArgumentException();
        };
    }

    static final BinaryOperator LPAREN = null;

    final int precedence;
    final boolean rightAssociative;

    BinaryOperator(int precedence, boolean rightAssociative) {
        this.precedence = precedence;
        this.rightAssociative = rightAssociative;
    }

    abstract double apply(double a, double b);
}
